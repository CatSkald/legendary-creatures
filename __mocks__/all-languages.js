import languages from "../src/i18n/languages";

exports.allLanguages = Object.entries(languages).map(
  ([lang, langProps]) => langProps,
);
