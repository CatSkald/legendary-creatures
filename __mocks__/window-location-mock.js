exports.mockWindowLocation = () => {
  const originalWindowLocation = window.location;

  beforeEach(() => {
    delete window.location;

    // mock window.location
    window.location = Object.defineProperties(
      {},
      {
        ...Object.getOwnPropertyDescriptors(originalWindowLocation),
        assign: {
          configurable: true,
          value: jest.fn(),
        },
      },
    );
  });

  afterEach(() => {
    // restore original `jsdom` location
    window.location = originalWindowLocation;
  });
};
