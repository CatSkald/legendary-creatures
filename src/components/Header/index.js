import {
  header__container,
  header,
  header__menu,
  header__languages,
} from "./index.module.scss";

import React, { useState } from "react";
import PropTypes from "prop-types";
import NavigationBar from "../NavigationBar";
import Languages from "../Languages";
import NavigationBarButtons from "../NavigationBarButtons";
import HamburgerButton from "../HamburgerButton";

const Header = (props) => {
  const [isMenuActive, toggleMenu] = useState(false);
  const [isLanguageSelectionActive, toggleLanguageSelection] = useState(false);

  function handleToggleMenu() {
    toggleLanguageSelection(false);
    toggleMenu(!isMenuActive);
  }

  function handleToggleLanguageSelection() {
    toggleMenu(false);
    toggleLanguageSelection(!isLanguageSelectionActive);
  }

  function handleLanguageSelected() {
    toggleMenu(false);
    toggleLanguageSelection(false);
  }

  return (
    <div className={header__container}>
      <div className={header}>
        <HamburgerButton
          handleClick={handleToggleMenu}
          isActive={isMenuActive}
        />
        <div className={header__menu}>
          <NavigationBar
            isActive={isMenuActive}
            handleToggleMenu={handleToggleMenu}
          />
        </div>
        <div className={header__languages}>
          <NavigationBarButtons
            selectedTheme={props.selectedTheme}
            handleToggleColorTheme={props.handleToggleColorTheme}
            isLanguageSelectionActive={isLanguageSelectionActive}
            handleToggleLanguageSelection={handleToggleLanguageSelection}
          />
          <Languages
            isActive={isLanguageSelectionActive}
            handleLanguageSelected={handleLanguageSelected}
            localizedLinks={props.localizedLinks}
          />
        </div>
      </div>
    </div>
  );
};

Header.propTypes = {
  handleToggleColorTheme: PropTypes.func.isRequired,
  selectedTheme: PropTypes.shape({
    name: PropTypes.string.isRequired,
    icon: PropTypes.string.isRequired,
    cssClassName: PropTypes.string.isRequired,
  }).isRequired,
};

export default Header;
