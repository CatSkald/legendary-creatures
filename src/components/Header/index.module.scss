.header__container {
  background-color: var(--menu-bg);
  border-width: 0;
  border-bottom-width: var(--border-width);
  border-color: var(--border-color);
  border-style: var(--border-style);
  position: sticky;
  top: 0;
  z-index: 9;
}

.header {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  flex-flow: column-reverse;
  align-items: flex-end;

  max-width: var(--container-width);
  margin: 0 auto;
  padding: var(--space-sm);
  position: relative;

  @media screen and (min-width: 768px /*--screen-medium*/) {
    flex-direction: row;
    align-items: center;
    padding: 0 var(--space);
  }
}

.header__menu {
  width: 100%;

  @media screen and (min-width: 768px /*--screen-medium*/) {
    left: var(--space-sm);
    margin-right: auto;
    width: auto;
  }
}

.header__languages {
  margin-left: var(--space);
}
