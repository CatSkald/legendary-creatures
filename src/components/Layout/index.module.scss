.layout__container {
  display: flex;
  min-height: 100vh;
  flex-direction: column;
}

.layout {
  display: flex;
  flex: 1;
  width: 100%;
}

.layout__content {
  width: 100%;
  max-width: var(--container-width);
  margin: 0 auto;
  padding: var(--space-sm);

  @media screen and (min-width: 768px /*--screen-medium*/) {
    padding: var(--space);
  }
}
