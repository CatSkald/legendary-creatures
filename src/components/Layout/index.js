import "../../styles/global.scss";
import "../../styles/theme-light.scss";
import "../../styles/theme-dark.scss";
import "../../styles/theme-dark-high-contrast.scss";
import {
  layout__container,
  layout,
  layout__content,
} from "./index.module.scss";

import React from "react";
import Header from "../Header";
import Footer from "../Footer";
import { loadUserSettings } from "../../utils/user-settings-helpers";

const LocaleContext = React.createContext();

const Layout = (props) => {
  var userSettings = loadUserSettings();
  const [selectedTheme, switchColorTheme] = React.useState(
    userSettings.selectedTheme,
  );

  function handleToggleColorTheme() {
    userSettings.toggleTheme();
    switchColorTheme(userSettings.selectedTheme);
  }

  return (
    <LocaleContext.Provider value={{ language: props.pageContext.language }}>
      <div
        className={`body ${layout__container} ${selectedTheme.cssClassName}`}
      >
        <Header
          selectedTheme={userSettings.selectedTheme}
          handleToggleColorTheme={handleToggleColorTheme}
          localizedLinks={props.pageContext.localizedLinks}
        />
        <section className={layout} role="main">
          <div className={layout__content}>{props.children}</div>
        </section>
        <Footer />
      </div>
    </LocaleContext.Provider>
  );
};

export { Layout, LocaleContext };
