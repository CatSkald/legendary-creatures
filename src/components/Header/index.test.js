import React from "react";
import { render } from "@testing-library/react";
import { LocaleContext } from "../Layout";

import Header from "../Header";

import { allLanguages } from "../../../__mocks__/all-languages";

jest.mock("../../i18n/translations/useTranslations");

//TODO validate props are being used properly
describe("Header", () => {
  const theme = Object.freeze({
    name: "Test Theme",
    icon: "â¨€",
    cssClassName: "theme-custom",
  });

  test.each(allLanguages)("renders correctly for language %s", (language) => {
    const { container } = render(
      <LocaleContext.Provider value={{ language }}>
        <Header handleToggleColorTheme={jest.fn()} selectedTheme={theme} />
      </LocaleContext.Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});
