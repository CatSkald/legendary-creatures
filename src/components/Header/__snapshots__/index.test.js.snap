// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly for language {
  name: 'Українська',
  code: 'uk',
  path: '/uk',
  dateFormat: 'DD.MM.YYYY'
} 1`] = `
<div>
  <div
    class="header__container"
  >
    <div
      class="header"
    >
      <button
        aria-label="Show menu"
        class="hamburgerMenu "
      >
        <span />
      </button>
      <div
        class="header__menu"
      >
        <nav
          class="navigationBar "
        >
          <a
            activeclassname="active"
            aria-label="Головна"
            class="navigationBar__link"
            href="/uk"
            hreflang="uk"
          >
            Головна
          </a>
          <a
            activeclassname="active"
            aria-label="Бібліотека істот"
            class="navigationBar__link"
            href="/uk/істоти"
            hreflang="uk"
          >
            Бібліотека істот
          </a>
          <a
            activeclassname="active"
            aria-label="Мітки"
            class="navigationBar__link"
            href="/uk/мітки"
            hreflang="uk"
          >
            Мітки
          </a>
          <a
            activeclassname="active"
            aria-label="Про проект"
            class="navigationBar__link"
            href="/uk/about"
            hreflang="uk"
          >
            Про проект
          </a>
          <a
            activeclassname="active"
            aria-label="Пошук"
            class="navigationBar__link"
            href="/uk/search"
            hreflang="uk"
          >
            Пошук
          </a>
        </nav>
      </div>
      <div
        class="header__languages"
      >
        <ul
          class="navigationBarButtons"
        >
          <li
            class="navigationBarButtons__button"
          >
            <button
              class="navigationBarButtons__language "
              title="Change language"
            >
              Українська
            </button>
          </li>
          <li
            class="navigationBarButtons__button"
          >
            <button
              aria-label="Change website color theme"
              title="Theme 'Test Theme', click to change"
            >
              <span
                class="navigationBarButtons__theme"
                role="img"
              >
                ⨀
              </span>
            </button>
          </li>
        </ul>
        <ul
          class="languageMenu "
        >
          <li
            class="languageMenu__language"
          >
            <a
              class="languageMenu__link "
              href="/"
              hreflang="en"
            >
              English
            </a>
          </li>
          <li
            class="languageMenu__language"
          >
            <a
              class="languageMenu__link "
              href="/ru"
              hreflang="ru"
            >
              Русский
            </a>
          </li>
          <li
            class="languageMenu__language"
          >
            <a
              class="languageMenu__link active"
              href="/uk"
              hreflang="uk"
            >
              Українська
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`Header renders correctly for language { name: 'English', code: 'en', path: '', dateFormat: 'DD/MM/YYYY' } 1`] = `
<div>
  <div
    class="header__container"
  >
    <div
      class="header"
    >
      <button
        aria-label="Show menu"
        class="hamburgerMenu "
      >
        <span />
      </button>
      <div
        class="header__menu"
      >
        <nav
          class="navigationBar "
        >
          <a
            activeclassname="active"
            aria-label="Home"
            class="navigationBar__link"
            href="/"
            hreflang="en"
          >
            Home
          </a>
          <a
            activeclassname="active"
            aria-label="Creatures"
            class="navigationBar__link"
            href="/creatures"
            hreflang="en"
          >
            Creatures
          </a>
          <a
            activeclassname="active"
            aria-label="Tags"
            class="navigationBar__link"
            href="/tags"
            hreflang="en"
          >
            Tags
          </a>
          <a
            activeclassname="active"
            aria-label="About"
            class="navigationBar__link"
            href="/about"
            hreflang="en"
          >
            About
          </a>
          <a
            activeclassname="active"
            aria-label="Search"
            class="navigationBar__link"
            href="/search"
            hreflang="en"
          >
            Search
          </a>
        </nav>
      </div>
      <div
        class="header__languages"
      >
        <ul
          class="navigationBarButtons"
        >
          <li
            class="navigationBarButtons__button"
          >
            <button
              class="navigationBarButtons__language "
              title="Change language"
            >
              English
            </button>
          </li>
          <li
            class="navigationBarButtons__button"
          >
            <button
              aria-label="Change website color theme"
              title="Theme 'Test Theme', click to change"
            >
              <span
                class="navigationBarButtons__theme"
                role="img"
              >
                ⨀
              </span>
            </button>
          </li>
        </ul>
        <ul
          class="languageMenu "
        >
          <li
            class="languageMenu__language"
          >
            <a
              class="languageMenu__link active"
              href="/"
              hreflang="en"
            >
              English
            </a>
          </li>
          <li
            class="languageMenu__language"
          >
            <a
              class="languageMenu__link "
              href="/ru"
              hreflang="ru"
            >
              Русский
            </a>
          </li>
          <li
            class="languageMenu__language"
          >
            <a
              class="languageMenu__link "
              href="/uk"
              hreflang="uk"
            >
              Українська
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`Header renders correctly for language { name: 'Русский', code: 'ru', path: '/ru', dateFormat: 'DD.MM.YYYY' } 1`] = `
<div>
  <div
    class="header__container"
  >
    <div
      class="header"
    >
      <button
        aria-label="Show menu"
        class="hamburgerMenu "
      >
        <span />
      </button>
      <div
        class="header__menu"
      >
        <nav
          class="navigationBar "
        >
          <a
            activeclassname="active"
            aria-label="Главная"
            class="navigationBar__link"
            href="/ru"
            hreflang="ru"
          >
            Главная
          </a>
          <a
            activeclassname="active"
            aria-label="Библиотека существ"
            class="navigationBar__link"
            href="/ru/существа"
            hreflang="ru"
          >
            Библиотека существ
          </a>
          <a
            activeclassname="active"
            aria-label="Метки"
            class="navigationBar__link"
            href="/ru/метки"
            hreflang="ru"
          >
            Метки
          </a>
          <a
            activeclassname="active"
            aria-label="О проекте"
            class="navigationBar__link"
            href="/ru/about"
            hreflang="ru"
          >
            О проекте
          </a>
          <a
            activeclassname="active"
            aria-label="Поиск"
            class="navigationBar__link"
            href="/ru/search"
            hreflang="ru"
          >
            Поиск
          </a>
        </nav>
      </div>
      <div
        class="header__languages"
      >
        <ul
          class="navigationBarButtons"
        >
          <li
            class="navigationBarButtons__button"
          >
            <button
              class="navigationBarButtons__language "
              title="Change language"
            >
              Русский
            </button>
          </li>
          <li
            class="navigationBarButtons__button"
          >
            <button
              aria-label="Change website color theme"
              title="Theme 'Test Theme', click to change"
            >
              <span
                class="navigationBarButtons__theme"
                role="img"
              >
                ⨀
              </span>
            </button>
          </li>
        </ul>
        <ul
          class="languageMenu "
        >
          <li
            class="languageMenu__language"
          >
            <a
              class="languageMenu__link "
              href="/"
              hreflang="en"
            >
              English
            </a>
          </li>
          <li
            class="languageMenu__language"
          >
            <a
              class="languageMenu__link active"
              href="/ru"
              hreflang="ru"
            >
              Русский
            </a>
          </li>
          <li
            class="languageMenu__language"
          >
            <a
              class="languageMenu__link "
              href="/uk"
              hreflang="uk"
            >
              Українська
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
